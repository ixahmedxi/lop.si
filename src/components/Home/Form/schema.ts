import * as yup from 'yup'

const blacklist = [
  'bit.ly',
  'cutt.ly',
  'shorturl.at',
  'rebrandly',
  'tinyurl',
  'tiny.cc',
  'raboninco',
  'is.gd'
]

export const schema = yup.object().shape({
  url: yup
    .string()
    .required()
    .url()
    .test('is-short', 'Url cannot be a short url', (value: string) => !value.includes('://lop.si'))
    .test('is-last-url', 'Url has been recently shortened', (value: string) => {
      const lastUrl = window.localStorage.getItem('last-url')
      return !(lastUrl === value)
    })
    .test('is-blacklisted', 'Url provided is in our blacklist', (value: string) => {
      let pass = true
      blacklist.forEach((item) => {
        if (value.includes(item)) {
          pass = false
        }
      })
      return pass
    })
})
