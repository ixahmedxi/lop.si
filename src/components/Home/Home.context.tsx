import { createContext, Dispatch, SetStateAction, useState } from 'react'

interface IHomeContext {
  id: string
  setId: Dispatch<SetStateAction<string>>
}

export const HomeContext = createContext<IHomeContext>({
  id: '',
  setId: () => null
})

export const HomeProvider: React.FC = ({ children }) => {
  const [id, setId] = useState('')
  return <HomeContext.Provider value={{ id, setId }}>{children}</HomeContext.Provider>
}
