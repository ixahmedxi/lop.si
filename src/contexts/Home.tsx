import { createContext, Dispatch, SetStateAction, useContext, useState } from 'react'

interface IHomeContext {
  id: string
  setId: Dispatch<SetStateAction<string>>
  isLoading: boolean
  setIsLoading: Dispatch<SetStateAction<boolean>>
}

export const HomeContext = createContext<IHomeContext>({
  id: '',
  setId: () => null,
  isLoading: false,
  setIsLoading: () => false
})

export const HomeProvider: React.FC = ({ children }) => {
  const [id, setId] = useState('')
  const [isLoading, setIsLoading] = useState(false)
  return (
    <HomeContext.Provider value={{ id, setId, isLoading, setIsLoading }}>
      {children}
    </HomeContext.Provider>
  )
}

export const useHomeContext = (): IHomeContext => useContext(HomeContext)
