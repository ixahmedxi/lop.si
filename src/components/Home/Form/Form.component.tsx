/* @jsx jsx */
import { useFirestore } from '@contexts/Firebase'
import { useHomeContext } from '@contexts/Home'
import { yupResolver } from '@hookform/resolvers'
import { useNeuBoxShadow } from '@hooks/useBoxShadow'
import { useState } from 'react'
import { useForm } from 'react-hook-form'
import { FiPaperclip } from 'react-icons/fi'
import { Box, Input, jsx, SxStyleProp } from 'theme-ui'
import { FormErrors } from './FormErrors/FormErrors.component'
import { schema } from './schema'
import { SubmitButton } from './SubmitButton/SubmitButton.component'

export const Form: React.FC = () => {
  const { setId } = useHomeContext()
  const { createOneUrl } = useFirestore()
  const [isLoading, setIsLoading] = useState(false)
  const { register, handleSubmit, errors, reset } = useForm<{ url: string }>({
    resolver: yupResolver(schema)
  })

  const styles: SxStyleProp = {
    wrapper: {
      width: ['100%', '100%', '100%', '70%'],
      m: '0 auto',
      pt: [3, 3, 5]
    },
    form: {
      display: 'grid',
      gridTemplateColumns: 'auto 1fr auto',
      placeItems: 'center',
      borderRadius: 100,
      p: '4px',
      border: '3px solid',
      borderColor: typeof errors.url !== 'undefined' ? '#e8505b' : 'transparent',
      transition: 'border-color 0.2s ease-out',
      ...useNeuBoxShadow(10, 20)
    },
    input: {
      outline: 'none',
      border: 'none',
      px: 3,
      height: [5, 5, 6],
      fontSize: [0, 0, 1],
      '::placeholder': {
        color: 'text',
        opacity: 0.5
      }
    },
    icon: {
      fontSize: [0, 0, 1],
      opacity: 0.5,
      ml: [3, 3, 5]
    }
  }

  const onFormSubmit = handleSubmit(async ({ url }) => {
    setIsLoading(true)
    const createdId = await createOneUrl(url)
    window.localStorage.setItem('last-url', url)
    setId(String(createdId))
    setIsLoading(false)
    reset()
  })

  return (
    <Box sx={styles.wrapper}>
      <form sx={styles.form} onSubmit={onFormSubmit} autoComplete="off" noValidate>
        <FiPaperclip sx={styles.icon} />
        <Input
          type="url"
          name="url"
          sx={styles.input}
          placeholder="Paste your url here..."
          ref={register}
        />
        <SubmitButton isLoading={isLoading} />
      </form>
      <FormErrors errors={errors} />
    </Box>
  )
}
