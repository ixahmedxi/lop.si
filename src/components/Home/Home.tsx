import { HomeProvider } from '@contexts/Home'
import { NextPage } from 'next'
import { Card } from './Card/Card.component'
import { Form } from './Form/Form.component'
import { Introduction } from './Introduction/Introduction.component'

const Home: NextPage = () => {
  return (
    <HomeProvider>
      <Introduction />
      <Form />
      <Card />
    </HomeProvider>
  )
}

export default Home
