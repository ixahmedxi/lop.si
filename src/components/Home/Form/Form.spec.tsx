import { HomeProvider } from '@contexts/Home'
import { render } from 'test-utils'
import { Form } from './Form.component'

describe('Form component', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <HomeProvider>
        <Form />
      </HomeProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
