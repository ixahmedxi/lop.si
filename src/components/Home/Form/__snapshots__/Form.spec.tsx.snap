// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component matches snapshot 1`] = `
.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  margin: 0 auto;
  padding-top: 32px;
}

@media screen and (min-width:576px) {
  .emotion-4 {
    width: 100%;
    padding-top: 32px;
  }
}

@media screen and (min-width:768px) {
  .emotion-4 {
    width: 100%;
    padding-top: 40px;
  }
}

@media screen and (min-width:992px) {
  .emotion-4 {
    width: 70%;
  }
}

.emotion-3 {
  display: grid;
  grid-template-columns: auto 1fr auto;
  place-items: center;
  border-radius: 100px;
  padding: 4px;
  border: 3px solid;
  border-color: transparent;
  -webkit-transition: none;
  transition: none;
  box-shadow: none;
}

.emotion-0 {
  color: var(--theme-ui-colors-text,#EEEEEE);
  -webkit-transition: color 0.2s ease-out;
  transition: color 0.2s ease-out;
  font-size: 12px;
  opacity: 0.5;
  margin-left: 24px;
}

@media screen and (min-width:576px) {
  .emotion-0 {
    font-size: 12px;
    margin-left: 24px;
  }
}

@media screen and (min-width:768px) {
  .emotion-0 {
    font-size: 16px;
    margin-left: 40px;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 16px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 4px;
  color: inherit;
  background-color: transparent;
  outline: none;
  border: none;
  padding-left: 24px;
  padding-right: 24px;
  height: 40px;
  font-size: 12px;
}

@media screen and (min-width:576px) {
  .emotion-1 {
    height: 40px;
    font-size: 12px;
  }
}

@media screen and (min-width:768px) {
  .emotion-1 {
    height: 48px;
    font-size: 16px;
  }
}

.emotion-1::-webkit-input-placeholder {
  color: var(--theme-ui-colors-text,#EEEEEE);
  opacity: 0.5;
  -webkit-transition: color 0.2s ease-out;
  transition: color 0.2s ease-out;
}

.emotion-1::-moz-placeholder {
  color: var(--theme-ui-colors-text,#EEEEEE);
  opacity: 0.5;
  -webkit-transition: color 0.2s ease-out;
  transition: color 0.2s ease-out;
}

.emotion-1:-ms-input-placeholder {
  color: var(--theme-ui-colors-text,#EEEEEE);
  opacity: 0.5;
  -webkit-transition: color 0.2s ease-out;
  transition: color 0.2s ease-out;
}

.emotion-1::placeholder {
  color: var(--theme-ui-colors-text,#EEEEEE);
  opacity: 0.5;
  -webkit-transition: color 0.2s ease-out;
  transition: color 0.2s ease-out;
}

.emotion-2 {
  height: 40px;
  width: 40px;
  outline: none;
  border: none;
  font-size: 16px;
  background-color: var(--theme-ui-colors-primary,#8093f2);
  color: var(--theme-ui-colors-background,#2B2D32);
  border-radius: 100px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-transition: none;
  transition: none;
  box-shadow: none;
}

@media screen and (min-width:576px) {
  .emotion-2 {
    height: 40px;
    width: 40px;
    font-size: 16px;
  }
}

@media screen and (min-width:768px) {
  .emotion-2 {
    height: 48px;
    width: 48px;
    font-size: 20px;
  }
}

<div>
  <div
    class="emotion-4"
    data-testid="form"
  >
    <form
      autocomplete="off"
      class="emotion-3"
      novalidate=""
    >
      <svg
        class="emotion-0"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.44 11.05l-9.19 9.19a6 6 0 0 1-8.49-8.49l9.19-9.19a4 4 0 0 1 5.66 5.66l-9.2 9.19a2 2 0 0 1-2.83-2.83l8.49-8.48"
        />
      </svg>
      <input
        class="emotion-1"
        name="url"
        placeholder="Paste your url here..."
        type="url"
        value=""
      />
      <button
        aria-label="create a short url"
        class="emotion-2"
        type="submit"
      >
        <svg
          data-testid="arrow-right"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
          <polyline
            points="12 5 19 12 12 19"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
